src/index.js

```diff
@@ -109,8 +109,15 @@ function optimist(fn) {
           return revertReducer(state, action);
       }
     }
-    let separated = separateState(state);
-    return baseReducer(separated.optimist, separated.innerState, action);
+
+    let {optimist, innerState} = separateState(state);
+    if (state) {
+      let nextState = fn(innerState, action);
+      if (nextState === innerState && !optimist.length) {
```

Comment:
I think there are two new problems here:

1. If `optimist.length > 0` we should skip straight to baseReducer, not also run `fn` first.
2. It looks like `fn` is potentially getting called twice here (if it applies a change). The `fn` should only ever be called once per action.